src/jobflow_remote/remote/host/remote.py

```diff
@@ -265,24 +271,56 @@ def is_connected(self) -> bool:
     def put(self, src, dst) -> None:
         self._check_connected()
 
-        self._execute_remote_func(self.connection.put, src, dst)
+        self._execute_remote_func(lambda host: host.connection.put, src, dst)
 
     def get(self, src, dst) -> None:
         self._check_connected()
 
-        self._execute_remote_func(self.connection.get, src, dst)
+        self._execute_remote_func(lambda host: host.connection.get, src, dst)
 
     def copy(self, src, dst) -> None:
         cmd = ["cp", str(src), str(dst)]
         self.execute(cmd)
 
-    def _execute_remote_func(self, remote_cmd, *args, **kwargs):
+    def _execute_remote_func(
+        self,
+        remote_cmd: Callable,
+        *args,
+        _connection_workdir: str | Path | None = None,
+        **kwargs,
+    ):
+        """
+        Execute a remote command using the host, with the option to
+        recreate the connection and retry the command in case of failure.
+
+        Parameters
+        ----------
+        remote_cmd
+            A function that takes the current RemoteHost instance as its argument
+            and returns the actual function to execute with the given args/kwargs
+        args
+            Positional arguments to pass to the remote command
+        _connection_workdir:
+            a path passed to Connection.cd to be used as working directory for the
+            command.
+        kwargs
+            Keyword arguments to pass to the remote command
+        """
+
+        def execute_cmd():
+            # create the function associated to the current connection
+            with ExitStack() as stack:
+                if _connection_workdir:
+                    stack.enter_context(self.connection.cd(_connection_workdir))
+                func = remote_cmd(self)
+                return func(*args, **kwargs)
```

Comment:
A little bit of why it is done this way and maybe why the initial way of calling this method (with self.connection.run for example instead of a lambda function that is then used here to get the command from self) would help :)

Reply:
added some explanation